packages/vscode-mdx/src/typescript-plugin.js

```diff
@@ -0,0 +1,107 @@
+ * @type {PluginModuleFactory} modules
+ */
+const init = (modules) => {
+  const {typescript: ts} = modules
+  return {
+    create(info) {
+      const enptySnapshot = ts.ScriptSnapshot.fromString('')
+      const getScriptSnapshot = info.languageServiceHost.getScriptSnapshot.bind(info.languageServiceHost)
+      const getScriptVersion = info.languageServiceHost.getScriptVersion.bind(info.languageServiceHost)
+      const getScriptKind = info.languageServiceHost.getScriptKind?.bind(info.languageServiceHost)
+      const getProjectVersion = info.languageServiceHost.getProjectVersion?.bind(info.languageServiceHost)
+      const configFileName = info.project.projectKind === ts.server.ProjectKind.Configured
+        ? info.project.getProjectName()
+        : undefined
+
+      info.languageServiceHost.getScriptSnapshot = (fileName) => {
+        if (!initialized && fileName.endsWith('.mdx')) {
+          return enptySnapshot
+        }
+        return getScriptSnapshot(fileName)
+      }
+      info.languageServiceHost.getScriptVersion = (fileName) => {
+        if (!initialized && fileName.endsWith('.mdx')) {
+          return 'initializing...'
+        }
+        return getScriptVersion(fileName)
+      }
+      if (getScriptKind) {
+        info.languageServiceHost.getScriptKind = (fileName) => {
+          if (!initialized && fileName.endsWith('.mdx')) {
+            return ts.ScriptKind.JSX // TODO
+          }
+          return getScriptKind(fileName)
+        }
+      }
+      if (getProjectVersion) {
+        info.languageServiceHost.getProjectVersion = () => {
+          if (!initialized) {
+            return getProjectVersion() + ',initializing...'
+          }
+          return getProjectVersion()
+        }
+      }
+
+      let initialized = false
+
+      loadPlugins(configFileName, ts).then((plugins) => {
+        const files = createFileProvider(
+          [getLanguageModule(ts, plugins)],
+          ts.sys.useCaseSensitiveFileNames,
+          (fileName) => {
+            const snapshot = getScriptSnapshot(fileName)
+            if (snapshot) {
+              files.updateSourceFile(
+                fileName,
+                resolveCommonLanguageId(fileName),
+                snapshot
+              )
+            } else {
+              files.deleteSourceFile(fileName)
+            }
+          }
+        )
+
+        decorateLanguageService(files, info.languageService, true)
+        decorateLanguageServiceHost(files, info.languageServiceHost, ts, ['.mdx'])
+
+        info.project.markAsDirty()
+        initialized = true
+      })
+
+      return info.languageService
+    },
+    getExternalFiles(project, updateLevel = 0) {
+      if (updateLevel >= 1 || !externalFiles.has(project)) {
+        const newFiles = searchExternalFiles(ts, project, ['.mdx'])
+        externalFiles.set(project, newFiles)
+        project.refreshDiagnostics()
+      }
+
+      return externalFiles.get(project) ?? []
+    }
+  }
+}
+
+exports = init
```

Comment:
Reassigning `exports` doesn’t do anything. That means this module is entirely unused.

I think you’re confusing the following TypeScript syntax:

```ts
export = init
```

which compiles to the following JavaScript:

```suggestion
module.exports = init
```